core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -15,71 +15,542 @@
+	BackCompat {
+		/// The unique substream where bidirectional communications happen.
+		substream: RegisteredProtocolSubstream<TMessage, TSubstream>,
+		/// Contains substreams which are being shut down.
+		shutdown: SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 4]>,
+	},
 
-	/// We are disabled. We close existing substreams and refuse incoming connections, but don't
-	/// shut down the entire handler.
-	Disabled,
+	/// Normal functionning. Contains the substreams that are open.
+	/// If we are in this state, we have sent a `CustomProtocolOpen` message to the outside.
+	Normal(PerProtocolNormalState<TMessage, TSubstream>),
+
+	/// We are disabled. Contains substreams that are being closed.
+	/// If we are in this state, either we have sent a `CustomProtocolClosed` message to the
+	/// outside or we have never sent any `CustomProtocolOpen` in the first place.
+	Disabled {
+		/// List of substreams to shut down.
+		shutdown: SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 6]>,
+
+		/// If true, we should reactivate the handler after all the substreams in `shutdown` have
+		/// been closed.
+		///
+		/// Since we don't want to mix old and new substreams, we wait for all old substreams to
+		/// be closed before opening any new one.
+		reenable: bool,
+	},
 
-	/// We are trying to shut down the existing node and thus should refuse any incoming
-	/// connection.
-	ShuttingDown,
+	/// We are trying to shut down the connection and thus should refuse any incoming connection.
+	/// Contains substreams that are being closed. Once all the substreams are closed, we close
+	/// the connection.
+	ShuttingDown(SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 6]>),
+
+	/// We sometimes temporarily switch to this state during processing. If we are in this state
+	/// at the beginning of a method, that means something bad happend in the source code.
+	Poisoned,
+}
+
+/// Normal functionning mode for a protocol.
+struct PerProtocolNormalState<TMessage, TSubstream> {
+	/// Optional substream that we opened.
+	outgoing_substream: Option<RegisteredProtocolSubstream<TMessage, TSubstream>>,
+
+	/// Substreams that have been opened by the remote. We are waiting for a packet from it.
+	incoming_substreams: SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 4]>,
+
+	/// For each request that has been sent to the remote, contains the substream where we
+	/// expect a response.
+	pending_response: SmallVec<[(u64, RegisteredProtocolSubstream<TMessage, TSubstream>); 4]>,
+
+	/// For each request received by the remote, contains the substream where to send back our
+	/// response. Once a response has been sent, the substream closes.
+	pending_send_back: SmallVec<[(u64, RegisteredProtocolSubstream<TMessage, TSubstream>); 4]>,
+
+	/// List of messages waiting for a substream to open in order to be sent.
+	pending_messages: SmallVec<[TMessage; 6]>,
```

Comment:
I wonder if there should be some back pressure mechanism if this grows a lot? My understanding is that the current version measures the send queue of the `RegisteredProtocolSubstream` and issues `CustomProtosHandlerOut::Clogged` events if becomes too large. The new approach instead requests substreams as necessary and buffers messages here. If the connection becomes congested would it be useful to issue a `Clogged` event based on the length of this buffer?

Reply:
In theory, opening a substream is an asynchronous operation, but in practice it's not and they open instantly. Therefore I'm not really worried about this queue growing very large.
Instead we're moving the problem to the multiplexing layer, which to me is a more appropriate layer to handle network congestion.

You're right that this pretty much makes the `Clogged` event useless, as it will never fire with the new system. This is why "remove the clogged event" is the next item on the list in #1692.

Reply:
Are substreams not made available to `CustomProtocolsHandler` only after protocol negotiation which requires at least a full round trip via multistream-select? Would that not take a while over a congested connection?

Reply:
Ah, you're right, I'm a bit confused with the hybrid approach.
I'll add back some clogging verification here.
In a future PR we can embed the message to send directly in the `RegisteredProtocolSubstream` so that sending it is part of the upgrade, which will indeed delegate the problem to libp2p.